### PR TITLE
[Vouchers] Fix Sales Tax Totals By Order report

### LINKED ISSUE #11364: [Vouchers, Reports] Taxes in reports are not correct, e.g. Sales Tax Totals By Order report
## Description

The values in the Sales Tax Totals By Order report are not correct. Other reports may be affected as well.
  - Discounts on taxes should be reflected in the tax column (= original tax minus tax discount).
  - Total without tax needs to be updated accordingly.
image

## Expected Behavior

Values in reports must be correct.

## Actual Behaviour

Values in reports are not correct.

## Steps to Reproduce

1. Place an order with a voucher and taxes excluded from price.
2. Render the report.
3. Compare the numbers of the report to the numbers of the order in the back office.

## Animated Gif/Screenshot

See above.

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #11364  

Values it the Sales Tax Totals By Order report are not taking into account voucher when displaying tax related columns. 

#### What should we test?

This is a little tricky to test, as as far as I know, we can't access all the data we need from the UI. When tax are included in the price, we record the tax portion of the voucher against the `voucher_adjustment` itself but it's not displayed anywhere. When tax are excluded from the price, we record the tax portion of the voucher in it's own `voucher_adjustment`, which is then displayed in the edit order page in the backoffice. 

Calculation to check : 
Total excluding tax = order total - (tax amount - voucher tax amount)
Tax = tax amount - voucher tax amount 

##### Tax excluded from price :
Place an order with a voucher :
- Generate Sales Tax Totals by Order report 
- Using the number in the edit order page in the back office check that Total Excluding Tax and Tax column are correct 

##### Tax included in price :
Voucher tax amount can be retrieved by running the following in a rails console (replace <order number> by the order number) :

```ruby
# Retrieve the order
o = Spree::Order.find_by number: "<order number>"
# Print tax amount
o.voucher_adjustments.first&.included_tax.to_f
```

Place an order with a voucher :
- Generate Sales Tax Totals by Order report 
- Using the number in the edit order page in the back office and the voucher tax amount check that Total Excluding Tax and Tax column are correct 

It's a bit more complicated if there are more than one tax rate in play, but this is covered by automated specs. 

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1 or DFC)
- [ ] Technical changes only
- [x] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/spree/adjustment.rb [hunk 1]
```diff
@@ -70,6 +70,7 @@ class Adjustment < ApplicationRecord
     scope :credit, -> { where('amount < 0') }
     scope :return_authorization, -> { where(originator_type: "Spree::ReturnAuthorization") }
     scope :voucher, -> { where(originator_type: "Voucher") }
+    scope :voucher_tax, -> { where(originator_type: "Voucher").where("label LIKE 'Tax%'") }
```
Comment (dacook) [inline app/models/spree/adjustment.rb:73]: In [update_tax_adjustment_for](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/voucher_adjustments_service.rb#L53) we're currently defining the conditions for finding a tax adjustment on an order, and this rule is much less specific. In particular it makes me uncomfortable that we have to partial match a label (what if someone manually creates an adjustment labeled "Tax"? Should we make it a reserved word?). 
Also uncomfortable using a `first` below.

I think perhaps the VoucherAdjustmentsService should have a method to find the order's voucher tax adjustment, which allows us to be more explicit about how many records may be found (zero or one for now, but maybe multiple in the future if we have multiple vouchers per order?)

What do you think, does that make sense?

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb [hunk 2]
```diff
@@ -104,9 +104,7 @@
     end
 
     it "generates the report" do
-      login_as admin
-      visit admin_reports_path
-      click_on "Sales Tax Totals By Order"
+      visit_sales_tax_totals_by_order
```
Comment (mkllnk) [inline spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:107]: Good idea but ideally we don't go the whole path in every spec. It's good to verify once that this path works but then we can navigate straight to the report page without going to the overview first. That saves some time.

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb [hunk 3]
```diff
@@ -153,6 +153,82 @@
...
+
+        # the enterprise fees can be known only when the user selects the variants
+        # we'll need to create them by calling recreate_all_fees!
+        order.recreate_all_fees!
+
+        order.update_totals_and_states
+        OrderWorkflow.new(order).complete!
+      end
+
+      pending "generates the report with the correct tax amount" do
```
Comment (mkllnk) [inline spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb]: Excellent test-driven dev.

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb [hunk 4]
```diff
@@ -153,6 +153,82 @@
...
+        expect(page).to have_button("Go")
+        click_on "Go"
+        expect(page.find("table.report__table thead tr").text).to have_content(table_header)
+
+        # Calculation details :
+        # Price excl tax = order.total - total_tax + voucher tax (negative number)
+        # Tax amount = Non adjusted tax - voucher tax
+        # Price incl tac = Price excl tax + non adjusted tax
+        expect(page.find("table.report__table tbody").text).to have_content([
+          "DistributorEnterpriseWithTax",
```
Comment (mkllnk) [inline spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:189]: Does this need to be a system spec? Could we test this on the lib level? It would be a lot faster and maybe more convenient to have the proper arrays already.

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (dacook) [reply app/models/spree/adjustment.rb:73]: 🤔 Hmm.. or would this be better modelled with standard ActiveModel features? I forget why we don't have a model for VoucherAdjustment..

Comment (mkllnk) [inline app/models/spree/adjustment.rb:73]: That should work for now but it's not ideal. I'm not sure what the ideal solution would be though. Maybe if we created voucher tax adjustments for each tax adjustment? Then we knew that it's voucher tax because it's coming from a voucher and adjusting a tax adjustment. Our of scope here though.

File: lib/reporting/reports/sales_tax/sales_tax_totals_by_order.rb [hunk 5]
```diff
@@ -159,7 +159,10 @@ def non_adjusted_tax_rate_total(query_result_row)
         end
 
         def voucher_tax_adjustment(order)
-          order.all_adjustments.voucher_tax.first&.amount || 0
+          included_tax_voucher_adjustment = order.voucher_adjustments.first&.included_tax || 0
```
Comment (dacook) [inline lib/reporting/reports/sales_tax/sales_tax_totals_by_order.rb:162]: I also found this part hard to understand. Eg `Adjustment` also has method `included_tax_total` which looks up any associated tax adjustments. It seems to be a complex web.. and probably I'm just not following it today.

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (mkllnk) [reply app/models/spree/adjustment.rb:73]: If this is a problem going forward then we should fix it soon though. There's also the issue of multiple tax rates, some inclusive and some exclusive. Related but doesn't necessarily require splitting the voucher tax adjustment... :thinking:

**Review by dacook — COMMENTED** [review]
This looks like it solves the problem. I wasn't able to understand the numbers but trust that you have that in hand.

However I had a thought about moving some logic to the VoucherAdjustmentsService.

**Review by mkllnk — APPROVED** [review]
Nice pull request. The code is looking good. I didn't check the numbers. I think that we still have to find a better model for voucher adjustments but that's out of scope here. (a bit worried that it will always be out of scope if we don't schedule it)

File: lib/reporting/reports/sales_tax/sales_tax_totals_by_order.rb (hunk 5 above)
Comment (rioug) [reply lib/reporting/reports/sales_tax/sales_tax_totals_by_order.rb:162]: `included_tax_total` uses adjustments with `originator_type` of `Spree::TaxRate` as you noted. In case of a voucher adjustment we store tax data in 2 different ways: 
- tax included in price, tax data is stored in `included_tax` in the voucher adjustment
- tax excluded in price, tax data is stored in it's own `voucher_adjustment` with a label that starts with "Tax", which is a bit clunky

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb [hunk 6]
```diff
@@ -153,6 +153,82 @@
...
+          "oc1",
+          "ORDER_NUMBER_1",
+          "GST Food",
+          "State",
+          "1.5 %",
+          "105.39", # 109.61 - 4.61 + 0.39
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:195]: I am not sure where you got your `115.00` from ? The spec above has the order total = 119.61 , so if you apply a $10 voucher  the order total becomes `109.61`.

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (rioug) [reply app/models/spree/adjustment.rb:73]: Yeah I think the the modelling of Voucher Tax adjustment isn't great at the moment and should probably be fixed ( Maikel raised similar concerns ) . I'll look into it a bit more, maybe we could use the `voucher_adjustment` as an `originator_type` for the adjustment that stores the tax part of the voucher when tax are excluded from price.

To answer your question, the label Tax is problematic, one could create a voucher called "Tax123" and it would break the current implementation. Currently, it is only possible to add one voucher to an order, at least via the UI, hence the use of `first` here. It's not restricted by the code. so technically you should be able to have more than one voucher per order.

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb (hunk 4 above)
Comment (rioug) [reply spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:189]: We could probably test this at the lib level. I am not really familiar with report, so it was easier to plug my scenario in the existing system spec. Do you think I should move the test out of the system spec ?

Comment (mkllnk) [reply spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:189]: > Do you think I should move the test out of the system spec ?

Yes, I think that would be good.

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (rioug) [reply app/models/spree/adjustment.rb:73]: @maikel @dacook What about using `AdjustmentMetadata` ?
We could use `fee_name: Tax, fee_type: "Voucher"` in the metadata associated with the `voucher_adjustment` describing the tax part. Then you can do something like : 
```ruby
order.voucher_adjustments.joins(:metadata).where(metadata: { fee_name: "Tax", fee_type: "Voucher" })
```
That way we still keep the voucher data as an order adjustment but we get rid of the ugly label match : `"label LIKE 'Tax%'"`

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (dacook) [reply spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb:195]: I think I was looking at this value: https://github.com/openfoodfoundation/openfoodnetwork/pull/11543/files/29a4bf88d7520b2e26615fa7e44a9a6a27c65ef7..a7bceb0b28a7ac863126ce59db15039bce4894a0#diff-ef69e93a7cc19d5540d0ef31a1153f01703706454194c29d2588cadffca68e11R120
But I see now that was total *excl.* tax. The total *incl.* tax is `116.73` though..

In any case, I'm thinking it would be good to arrange our tests in a way that show what the values mean. In this situation I would probably add a comment to each value. 
But I wonder if we could construct a helper to express these expections better. Something like:

```
food_row = row_with_content("DistributorEnterpriseWithTax oc1 ORDER_NUMBER_1 GST Food")
expect(food_row).to have_column_value("Total excl. Tax ($)": "115.0")
```

Hmm that's a bit wordy.. maybe I'll stick with my comment idea ;)

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (dacook) [reply app/models/spree/adjustment.rb:73]: I'm not familiar with AdjustmentMetadata yet so can't comment yet.. Maikel if you can comment please do, otherwise let me know and I'll look into it.

Comment (rioug) [reply app/models/spree/adjustment.rb:73]: As far as I can see it's only used in reports : https://github.com/openfoodfoundation/openfoodnetwork/blob/693d5bad4aa610e13b3b366172ebd727710207a6/lib/reporting/reports/enterprise_fee_summary/scope.rb#L69C1-L76

Comment (mkllnk) [reply app/models/spree/adjustment.rb:73]: Yes, it's another join but if we need to do that then we should. It's better then relying on the label.

File: spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb [hunk 7]
```diff
@@ -467,6 +462,12 @@
     end
   end
 
+  def visit_sales_tax_totals_by_order
+    login_as admin
+    visit admin_reports_path
+    click_on "Sales Tax Totals By Order"
+  end
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb]: Done here https://github.com/openfoodfoundation/openfoodnetwork/pull/11543/commits/19fc1fab8c2a12b4484c8723e1b8883273862f92

File: app/models/spree/adjustment.rb (hunk 1 above)
Comment (rioug) [reply app/models/spree/adjustment.rb]: Fixed here : https://github.com/openfoodfoundation/openfoodnetwork/pull/11543/commits/03ce39d5c536fc0db41951dd1d185b2c423c883e  no more hacky `LIKE` matching

File: app/services/voucher_adjustments_service.rb [hunk 8]
```diff
@@ -65,6 +84,14 @@ def update_tax_adjustment_for(adjustment, tax_amount)
...
     tax_adjustment.amount = tax_amount
     tax_adjustment.save
+
+    # Add metada so we know which voucher adjustment is Tax related
+    AdjustmentMetadata.create(
+      adjustment: tax_adjustment,
+      enterprise: adjustment.originator.enterprise,
+      fee_name: "Tax",
+      fee_type: "Voucher"
+    )
```
Comment (mkllnk) [inline app/services/voucher_adjustments_service.rb:94]: Nice. I would probably build the data structure first and then call save only once. Rails may put it all in one transaction then.
```suggestion
    tax_adjustment.amount = tax_amount

    # Add metada so we know which voucher adjustment is Tax related
    tax_adjustment.metadata ||= AdjustmentMetadata.new(
      enterprise: adjustment.originator.enterprise,
      fee_name: "Tax",
      fee_type: "Voucher"
    )

    tax_adjustment.save
```

I also did an optional assignment here in case we are updating an existing adjustment which has metadata already. The metadata shouldn't change, I think, and therefore we don't need to update it.

File: app/models/voucher.rb [hunk 9]
```diff
@@ -13,6 +13,9 @@ class Voucher < ApplicationRecord
            dependent: :nullify
 
   validates :code, presence: true, uniqueness: { scope: :enterprise_id }
+  # We store the tax portion of a voucher in a separate adjustment with a label starting by "Tax"
+  # when tax are exluded from price. To avoid any issue we a code stating by Tax in not allowed
+  validates :code, format: { without: /\ATax/ }
```
Comment (mkllnk) [inline app/models/voucher.rb:18]: Is this still needed?

File: app/services/voucher_adjustments_service.rb [hunk 10]
```diff
@@ -5,6 +5,10 @@ def initialize(order)
     @order = order
   end
 
+  # The tax part of the voucher is stored as explained below:
+  # * tax included in price: included_tax field of the voucher adjustment
+  # * tax exckuded from price: as an extra voucher adjustment, with label starting by "Tax "
+  #
```
Comment (mkllnk) [inline app/services/voucher_adjustments_service.rb:11]:
```suggestion
```

File: config/locales/en.yml [hunk 11]
```diff
@@ -117,6 +117,10 @@ en:
             using_producer_stock_settings_but_count_on_hand_set: "must be blank because using producer stock settings"
             on_demand_but_count_on_hand_set: "must be blank if on demand"
             limited_stock_but_no_count_on_hand: "must be specified because forcing limited stock"
+        voucher:
+          attributes:
+            code:
+              invalid: "can't start with 'Tax'"
```
Comment (mkllnk) [inline config/locales/en.yml:123]: Still needed?

File: app/services/voucher_adjustments_service.rb (hunk 8 above)
Comment (rioug) [reply app/services/voucher_adjustments_service.rb:94]: Nice one, That's correct the metadata doesn't need to change even if the adjustment is updated.

File: app/models/voucher.rb (hunk 9 above)
Comment (rioug) [reply app/models/voucher.rb:18]: Not really, it might lead to some funny display combination but it won't break the code. I'll remove it

File: spec/lib/reports/sales_tax_totals_by_order_spec.rb [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        create(:voucher_flat_rate, code: 'some_code', enterprise: order.distributor, amount: 10)
+      end
+
+      it "returns the total exluding tax and indcluding voucher tax discount" do
+        add_voucher(order, voucher)
+
+        total = report.total_excl_tax(query_row)
+
+        # order.total - order.total_tax + voucher tax
+        expect(total).to eq(103.3 - 3.3 + 0.29)
```
Comment (dacook) [inline spec/lib/reports/sales_tax_totals_by_order_spec.rb:156]: At first I think I understand. The order is 113.3 including tax. But the voucher discounts the order by 10, meaning the customer only pays 103.3. Tax is only applicable to what was paid, so only now there is only 3.01 tax (3.3-0.29). 

Hmm I could be wrong, but I think this comment is misleading. I think the model method `order.total_tax` would return the discounted `3.01` (just like `order.total` returns the discounted `103.3`). 
Is this more correct? (I'm not sure, the numbers still seem wrong..) 

```suggestion
        # discounted order total - discounted order tax
        expect(total).to eq((113.3 - 10) - (3.3 - 0.29))
```

File: spec/lib/reports/sales_tax_totals_by_order_spec.rb [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+      it "returns expected totals" do
+        expected_summary = {
+          total_excl_tax: 110.00,
+          tax: 3.3,
+          total_incl_tax: 113.30,
+          first_name: "cfname",
+          last_name: "clname",
+          code: "ABC123",
+          email: "order1@example.com"
+        }
```
Comment (dacook) [inline spec/lib/reports/sales_tax_totals_by_order_spec.rb:215]: I would embed the expected values down below in the `expect` statement at the end, I think it reads better that way.

File: app/services/voucher_adjustments_service.rb [hunk 14]
```diff
@@ -85,6 +83,14 @@ def update_tax_adjustment_for(adjustment, tax_amount)
     # Update the amount if tax adjustment already exist, create if not
     tax_adjustment = @order.adjustments.find_or_initialize_by(adjustment_attributes)
     tax_adjustment.amount = tax_amount
+
+    # Add metada so we know which voucher adjustment is Tax related
+    tax_adjustment.metadata ||= AdjustmentMetadata.new(
+      enterprise: adjustment.originator.enterprise,
```
Comment (dacook) [inline app/services/voucher_adjustments_service.rb:89]: This only gets set once, when the tax adjustment is initialised, so I wondered if this value could get out of date. But I don't think so: I expect the order/voucher enterprise can't change.

File: spec/services/voucher_adjustments_service_spec.rb [hunk 15]
```diff
@@ -342,6 +348,12 @@
         included_tax: 0,
         amount: 0.5
       )
+      AdjustmentMetadata.create(
+        adjustment: tax_adjustment,
+        enterprise: tax_adjustment.originator.enterprise,
+        fee_name: "Tax",
+        fee_type: "Voucher"
+      )
```
Comment (dacook) [inline spec/services/voucher_adjustments_service_spec.rb:356]: 👀 This is now quite a bit of manual set up duplicated from the application code..

**Review by dacook — CHANGES_REQUESTED** [review]
This looks good, and it's great to see more unit tests on the report! I was a little cautious about removing the system spec, but I think it's covered well, and better to do it this way (we only need the browser to test that the data is rendered in a table, we don't need it to test how data is processed in the backend).

I made a few other comments as I attempted to comprehend it. If they don't make sense you can ignore them. 

`±`Also I think I saw an opportunity to DRY the spec, which would be worth changing.

File: spec/lib/reports/sales_tax_totals_by_order_spec.rb (hunk 12 above)
Comment (rioug) [reply spec/lib/reports/sales_tax_totals_by_order_spec.rb:156]: You are right it's a bit misleading, it seems to make sense when I wrote it but not so much now.
FYI `order.total_tax` doesn't returned the discounted tax because :
1. voucher tax part isn't always modelled as an adjustment
2. when modelled as an adjustment it's not considered a "Tax adjustment"

I have been a bit reluctant to fix it, because I am not sure what the implication would be in other part of OFN.

File: spec/lib/reports/sales_tax_totals_by_order_spec.rb [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+    end
+
+    # TODO see if we can simplify setup
+    # sould be able to only create the adjustment we need, or maybe mock them ?
+    it "returns tax amount filtered by tax rate in query_row" do
+      OrderWorkflow.new(order).complete!
+
+      filtered_tax_total = report.filtered_tax_rate_total(query_row)
+
+      expect(filtered_tax_total).not_to eq(order.total_tax)
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/lib/reports/sales_tax_totals_by_order_spec.rb:82]: We can specify the value of `order.total_tax` but then we would have to explain where the number is coming from. As it's, I feel like the code is self documenting, we don't really care what the value is we just want to make sure it's not the same as the filtered value.

File: spec/services/voucher_adjustments_service_spec.rb (hunk 15 above)
Comment (rioug) [reply spec/services/voucher_adjustments_service_spec.rb]: I fix that, thanks for prompting me to re visit this :pray:

**Review by dacook — APPROVED** [review]
Wonderful 👍

**Review by mkllnk — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @rioug,
Thanks for working on this important fix! :muscle: 

### Testing the Sales Tax Totals By Order report

I have first reproduced the exact same order as in the issue. We can now see, that the order in the back office and the Tax Totals Report By Order are matching well.
image

I then tested different scenarios with tax incl./excl. and with percentage vouchers. The numbers are all in coherence with the numbers in the back office, but I found an issue with percentage vouchers and tax excl. from prices: see [wishlist #476](https://github.com/openfoodfoundation/wishlist/issues/476) - it is out of scope for this one here though.

image

### Testing the other reports
Examples for No. 1 - 10 below.

| No. | Report | Result | Comment |
| - | - | - | - |
| 1 | Tax Types | vouchers not taken into account | :x: |
| 2 | Tax Rates | vouchers not taken into account | :x: |
| 3 | Sales Tax Total by Producer | vouchers not taken into account | probably ok :question: |
| - | Sales Tax Total by Order | vouchers taken into account | :heavy_check_mark: |
| - | Bulk Co-op Customer Payments | vouchers taken into account | :heavy_check_mark: |
| - | Payments By Type | vouchers taken into account | :heavy_check_mark: |
| - | Itemised Payment Totals | vouchers taken into account | :heavy_check_mark: |
| - | Payment Totals | vouchers taken into account | :heavy_check_mark: |
| - | Order Cycle Customer Totals | vouchers taken into account | :heavy_check_mark: |
| - | Customers | vouchers taken into account | :heavy_check_mark: |
| 4 | Enterprise Fees Summary | vouchers not taken into account | probably ok :question: |
| 5 | Enterprise Fees With Tax Report By Order | vouchers not taken into account | :x: |
| 6 | Enterprise Fees With Tax Report By Producer | vouchers not taken into account | :x: |
| - | Payment Methods Report | vouchers taken into account | :heavy_check_mark: |
| - | Delivery Report | vouchers taken into account | :heavy_check_mark: |
| 7 | Xero Summary | vouchers not taken into account | :x: |
| 7 | Xero Detailed | vouchers not taken into account | :x: |
| 8 | Pack By Customer | vouchers not taken into account | probably ok :question: |
| 9 | Pack By Supplier | vouchers not taken into account | probably ok :question: |
| 10 | Pack By Product | vouchers not taken into account | probably ok :question: |
| 11 | Revenues by Hub | vouchers not taken into account | :x: |

These are the correct values (including the voucher discount):
image

1. Subtract 10 % voucher:
image

2. Apply voucher not only to Total excl. tax but to Tax and Total tax as well:
image

3. Subtract 10 % voucher:
image

4. Subtract 10 % voucher:
image

5. Subtract 10 % voucher:
image

6. Subtract 10 % voucher:
image

7. Subtract 10 % voucher:
image

8. Subtract 10 % voucher:
image

9. Subtract 10 % voucher:
image

10. Subtract 10 % voucher:
image

11. Apply voucher not only to Total excl. tax but to Total tax as well:
image

## Summary
- This PR solves the incorrect tax calculation of #11364. :+1: 
- There are quite a few other reports which are affected by vouchers as well. I will ask on Slack for a prioritization.
- I couldn't spot any regressions so this one is ready for merging! :partying_face: 

Thanks again! :wink:
